src/Frame/Frame.cc

```diff
@@ -2334,4 +2329,221 @@ bool Frame::GetDownsampledRasterData(
+    // Prevent deleting the Frame while this task is not finished yet
+    std::shared_lock lock(GetActiveTaskMutex());
+
+    // Get vector field settings
+    int file_id = settings.file_id;
+    int mip = settings.smoothing_factor;
+    bool fractional = settings.fractional;
+    float threshold = (float)settings.threshold;
+    CARTA::CompressionType compression_type = settings.compression_type;
+    float compression_quality = settings.compression_quality;
+    int stokes_intensity = settings.stokes_intensity;
+    int stokes_angle = settings.stokes_angle;
+    double q_error = settings.q_error;
+    double u_error = settings.u_error;
+    bool debiasing = settings.debiasing;
+    if (!debiasing) {
+        q_error = u_error = 0;
+    }
+
+    // Set response messages
+    auto response =
+        Message::VectorOverlayTileData(file_id, _z_index, stokes_intensity, stokes_angle, compression_type, compression_quality);
+    auto* tile_pi = response.add_intensity_tiles();
+    auto* tile_pa = response.add_angle_tiles();
+
+    // Get tiles
+    std::vector<Tile> tiles;
+    GetTiles(_width, _height, mip, tiles);
+
+    // Set requirements
+    bool calculate_pi(stokes_intensity == 1), calculate_pa(stokes_angle == 1);
+    bool current_stokes_as_pi(stokes_intensity == 0), current_stokes_as_pa(stokes_angle == 0);
+
+    // Initialize stokes maps for their flags, indices and data
+    std::unordered_map<std::string, bool> stokes_flag{{"I", false}, {"Q", false}, {"U", false}};
+    std::unordered_map<std::string, int> stokes_indices{{"I", -1}, {"Q", -1}, {"U", -1}};
+    std::unordered_map<std::string, std::vector<float>> stokes_data{
+        {"I", std::vector<float>()}, {"Q", std::vector<float>()}, {"U", std::vector<float>()}};
+
+    // Set stokes flags and get their indices
+    stokes_flag["I"] = (fractional || !std::isnan(threshold));
+    stokes_flag["Q"] = stokes_flag["U"] = (calculate_pi || calculate_pa);
+    for (auto one : stokes_flag) {
+        std::string stokes = one.first;
+        if (stokes_flag[stokes] && !GetStokesTypeIndex(stokes + "x", stokes_indices[stokes])) {
+            return false;
+        }
+    }
+
+    // Get image tiles data
+    for (int i = 0; i < tiles.size(); ++i) {
+        auto& tile = tiles[i];
+        auto bounds = GetImageBounds(tile, _width, _height, mip);
+
+        // Get downsampled raster tile data
+        int width, height;
+        for (auto one : stokes_flag) {
+            std::string stokes = one.first;
+            if (stokes_flag[stokes] &&
+                !GetDownsampledRasterData(stokes_data[stokes], width, height, _z_index, stokes_indices[stokes], bounds, mip)) {
+                return false;
+            }
+        }
+
+        // Calculate the current stokes as polarized intensity or polarized angle
+        std::vector<float> current_stokes_data;
+        if (current_stokes_as_pi || current_stokes_as_pa) {
```

Comment:
There are a lot of `if` statements from here.  Can we put all the `if (current_stokes_as_*)` stuff in this block, or even use `Frame::FillDownsampledData` for both cases (no stokes or current stokes)?

And then have `if (calculate_pi)` and `if (calculate_pa)` blocks.  I think you could even move the calculations to VectorField, something like `if (calculate_pi) { CalculatePolarizedIntensity(tile_pi, tiles[i], stokes_data, settings, width, height); }` which calculates pi, applies transforms, and fills tile data.  Same for calculate_pa.

My concern, as we talked about at the F2F, is that the backend has very large classes (including Frame) and we should move as much code as possible to other files that do the work once we have the data.